### DISCUSSION TITLE
Helm - Quoting env vars that have special chars in values

### DISCUSSION THREAD

**Discussion by sbalmos** [vectordotdev/vector#17341] [Q&A]
I admit right off the start this is really more generally a Helm, YAML, and k8s ConfigMap question. But I can't tell if this almost might be an issue with how env vars are substituted into the Vector config before processing.

I have passwords that contain special characters, namely sometimes single quotes and double quotes (despite loud screaming from me to the password generator). I for the life of me cannot get a Helm quoting escape sequence that would properly preserve wrapping the password field with the needed double quotes, single quotes, etc.

Right now, say I have a Kafka `password` field, it's written as such:
`password: ${PASS}`

At startup, this substitutes in a string with a single quote, which then in Vector causes the config parser to panic, and startup fails.

I've tried `password: "${FOO}"`, which by the time Helm installs the generated ConfigMap, the double quotes have been removed. This is a well-known "feature" of Helm's `toYaml` marshalling. Also tried `password: '${FOO}'`, same results.

For completeness, I've also tried the other YAML escaping variations, with no success:

- `password: '${FOO}'` generates `password: ${FOO}`
- `password: "${FOO}"` generates `password: ${FOO}`
- `password: "'${FOO}'"` generates `password: '''${FOO}'''` (literal wrapping single-quotes in the password value. wrong password)
- `password: '"${FOO}"'` generates `password: '"${FOO}"'` (literal wrapping double-quotes in the password value, wrong password)
- `password: "\"${FOO}\""` generates `password: '"${FOO}"'` (same as above)

showing the examples of multi-line block quotes (e.g. `password: |`, `password: |-`, and `password: >`) screws up formatting in here. But they also vary from prepending the leading spaces into the password value, causing incorrect password, or having no effect at all.

Any other ideas or suggestions?

**Comment by spencergilbert** [comment]
👋 @sbalmos,

Appreciate the detail that went into this, as well as the related issue that you opened. I definitely think this is a problem with how Vector is currently handling environment variables, and linked an older issue that I believe is related.

I think we can discuss and work towards a solution on the GitHub issue(s). Thanks again.
